### ISSUE TITLE
Run command/script after stack startup?

### ISSUE THREAD

**Issue by strarsis** [play-with-docker/play-with-docker#154]
Using the stack GET parameter it is possible to let play with docker launch a stack automatically.
Is it also possible to run a command after the stack has been launched that executes commands inside the containers using docker-compose exec?

**Comment by xetorthio** [comment]
Right  now not automatically. But this is something we want to add soon!

**Comment by strarsis** [comment]
And an additional request which is also related:
Sometimes an app has to be configured to expect the right hostname + port (e.g. a WordPress site for a demo). Because the docker container itself knows only its own internal hostname, configuration from inside wouldn't work correctly - the actual publicly reachable hostname has to be determined from inside the PWD/shell on which the stack has been started. - A situation where running a command for setting things up also comes in handy.

**Comment by xetorthio** [comment]
Yes. We thought about injecting this to the container as an environment
variable.

**Comment by stealthybox** [comment]
`PWD_IP_ADDRESS` is already set in the d-in-d shells.
(not sure if this is set in the stack auto-exec environments)

unfortunately, docker stacks don't allow you to reoplace characters like this `${PWD_IP_ADDRESS//./-}`
:(
(throws a syntax error in validation)

I propose whichever ENV execs the `docker deploy` should have these vars:
```bash
PWD_IP_ADDRESS="10.0.82.3"
PWD_FQDN_IP_PREFIX="pwd10-0-82-3-"
PWD_FQDN_HOST_SUFFIX=".host1.labs.play-with-docker.com"
```

This would allow you to interpolate `http://pwd10-0-82-3-8080.host1.labs.play-with-docker.com` into the auto-stack files like this:
```yaml
version: '3.1'
services:
  my-app:
    image: alpine
    ports:
    - 8080:80
    environment:
      web_host: http://${PWD_FQDN_IP_PREFIX}8080${PWD_FQDN_HOST_SUFFIX}/
    entrypoint: echo ${web_host}
```

Probably best to leave container's environments alone unless the stack specifies the configuration.

(sorry for scope creep on this issue -- we might want to move this somewhere)

**Comment by marcosnils** [comment]
> Probably best to leave container's environments alone unless the stack specifies the configuration.

We're not parsing the stack file actually, so we don't know if it containers any env variables. Injecting the variables in the running container actually makes a lot of sense

**Comment by stealthybox** [comment]
@marcosnils you can interpolate env vars into a stack file 👍 
no need for custom parsing.

**Comment by strarsis** [comment]
@stealthybox: Is there an example? I want to interpolate the stdout of hostname command so I can use the docker hostname inside the containers.

**Comment by stealthybox** [comment]
Sure, here's a quick example of host shell variable interpolation:
https://github.com/adamcaron/flip/blob/master/dev-stack.yml#L6
That repo uses a shell script and a env/source file to parameterize the dev-workflow with the stack

You'll probably want to use the shellvar version of hostname instead of the command:
`${HOSTNAME}` vs `$(hostname)`

If you're trying to use container specific variables, you actually don't want to docker host's variable interpolation.

A good place to set options with container specific shellvars would be in the `entrypoint`.
You can escape the stackfile interpolation by using a double dollar sign: `$$HOSTNAME` <-- this will be `$HOSTNAME` inside of the container if used within a shell script in the `entrypoint`:
```yaml
  lb:
    image: someloadbalancer
    entrypoint:
    - sh
    - -c
    - server-binary --fqdn=$$HOSTNAME
```

See my docker-in-docker example.
It uses some useful nested interpolation for the $token value:
https://github.com/stealthybox/docker-demo/blob/master/swarm/docker-compose.yaml#L13-L29

**Comment by stealthybox** [comment]
@strarsis ^ lmk if that's helpful

back on topic:
these interpolation features can be useful for generic, declarative configuration of Play-with-docker stacks, because we can load the values of the interface in the execution environment for the stack.

**Comment by strarsis** [comment]
After setting up an HTTP server, how to access the node normally with a web browser?
Prepend the $HOSTNAME to something? It would be nice to have the URL shown per node.

**Comment by marcosnils** [comment]
> @marcosnils you can interpolate env vars into a stack file

@stealthybox yup, I'm aware about this. Some time in the conversation I forgot that we were talking about stacks. You can actually use interpolation variables there as well as the stack gets executed in the same container that gets created after that.

**Comment by strarsis** [comment]
@marcosnils, @xetorthio, @stealthybox: Any news on this? :)

**Comment by marcosnils** [comment]
@strarsis the env variabe thing has been added already. Regarding the specific functionally which this issue refers we're not sure how the API would look regarding executing a command once compose finishes. Why not creating another service with `restart-condition: none` like `swarm-exec` does?.

**Comment by strarsis** [comment]
So I want to use the hostname + port as docker-compose env so e.g. a WordPress site will correctly adjust the URLs - without the visitor would be redirected to an incorrect URL, the site won't work at all then.
With this new feature added, can I now do this or emulate this somehow?

**Comment by stealthybox** [comment]
@marcosnils I'm not sure it's the right approach but you can use extension objects with Docker Compose v3.4 now:
https://github.com/docker/cli/pull/452
```yaml
x-pwd-post-deploy:
  command:
    - echo
    - $${HOSTNAME}
```

I think your solution of specifying a one-shot service with `restart-condition: none` is a good one.

One thing I did notice is that the `PWD_IP_ADDRESS` environment variable is no longer injected into the d-in-d.
I think this is the relevant module:
https://github.com/play-with-docker/play-with-docker/blob/master/docker/docker.go#L303

There's still remnants of it in other places:
- https://github.com/play-with-docker/play-with-docker/blob/master/pwd/types/session.go#L20
- https://github.com/play-with-docker/play-with-docker/blob/master/dockerfiles/dind/Dockerfile.dind#L53
  ^ this one is interesting because it's no longer in the `daemon.json`

**Comment by strarsis** [comment]
@stealthybox: This new feature (which I can use for injecting outside environment variables into container as I intended?) - is the Docker (CLI/Composer) version supporting it used by current live play-with-docker?

**Comment by marcosnils** [comment]
> @marcosnils I'm not sure it's the right approach but you can use extension objects with Docker Compose v3.4 now:
> docker/cli#452

@stealthybox yeah, I've seen this approach but that'd require to use some custom features that only work in newer versions of compose. I'd prefer to avoid doing things like that.


> I think your solution of specifying a one-shot service with restart-condition: none is a good one.

Yup, I think that's the way that everyone does it currently.



> One thing I did notice is that the PWD_IP_ADDRESS environment variable is no longer injected into the d-in-d.
> I think this is the relevant module:

That PWD_IP_ADDRESS wasn't the container address but the IP of the PWD container in the session because we used things like internal DNS to resolve things. We've decided to remove that as it's not necessary anymore that's why we removed it. We can't inject the IP address of the container as an env variable because there's no way of knowing it before the container is created.

**Comment by marcosnils** [comment]
> So I want to use the hostname + port as docker-compose env so e.g. a WordPress site will correctly adjust the URLs - without the visitor would be redirected to an incorrect URL, the site won't work at all then.
> With this new feature added, can I now do this or emulate this somehow?

Can you please provide a stack example of what you're trying to achieve so I can better understand if it can be sorted out with what's already there?. I believe the "Exec commands after stack is up" objective is already covered with the `swarm-exec` strategy I've described before.

**Comment by strarsis** [comment]
@marcosnils:
So the docker-compose setup is upped, then a setup.sh script should be executed inside the container which sets up WordPress. Very important for the WordPress setup is the host and port for the proper base URL - otherwise the WordPress site will not work at all.

**Comment by marcosnils** [comment]
@strarsis I don't quite understand what you mean. We've deployed wordpress in PWD successfully before without having to specify any host. Have you tried this?. http://play-with-docker.com/?stack=/wordpress/latest/stack.yml

**Comment by strarsis** [comment]
@marcosnils: `wp core install` requires `--url` parameter, it doesn't seem to be optional like in the interactive WordPress setup which seems to autodiscover it correctly.

**Comment by strarsis** [comment]
@marcosnils: I just have discussed this issue in wp-cli/core-command#52:

> $_SERVER['HTTP_HOST'] is only present in web requests, so you'll always need to specify --url=<url> for CLI installation requests.

So for **automatically** setting up a WordPress installation one requires the host + port in the setup.sh script so it can pass it using the `--url` parameter.

**Comment by marcosnils** [comment]
@strarsis in that case you'll have to do some bash magic to send that flag in the command line
PWD Url's are composed as follows: http://ip<hyphen-ip>-<session_id>-<port>.direct.labs.play-with-docker.com (i.e: http://ip-2-135-3-b8ir6vbg5vr00095iil0-8080.direct.labs.play-with-docker.com).

The session_id can be obtained from the `SESSION_ID` env variable. So it's basically getting the ip address of the `eth0` interface and replacing `dots` for `hyphens`

**Comment by strarsis** [comment]
@marcosnils: Thanks!
Are there examples about how to use `swarm-exec` with play-with-docker? 
Do I have to add a `swarm-exec` field to the `docker-compose.yml` of the project?

**Comment by virtimus** [comment]
Currently it's rather eth1 that works - nevertheless I'm not able to get it automatically as it is available on host and not on containers. 
So - getting url part is not feasible.
It would be good to have some additional get parameters like "hostCommandBefore" "hostCommandAfter"

**Comment by strarsis** [comment]
@marcosnils: 
https://github.com/docker/compose/issues/5049

**Comment by marcosnils** [comment]
@strarsis ???.

**Comment by strarsis** [comment]
````
ERROR: The Compose file './docker-compose.yml' is invalid because:                                                                     Unsupported config option for services.wordpress: 'x-pwd-post-deploy'
````

**Comment by stealthybox** [comment]
@strarsis you need to set
```
version:
```
at the top of your compose yaml >= `'3.4'` in order to use extensions

I don't know why you linked that other issue.
It seems very unrelated.

**Comment by strarsis** [comment]
@stealthybox: docker-compose uses `version: '3.5'` now, but still the same error.
